KAFKA-10199: Enable state updater by default

cadonna opened the pull request:
Now that the implementation for the state updater is done, we can enable it by default.

This PR enables the state updater by default and fixes code that made assumptions that are not true when the state updater is enabled (mainly tests).



### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

cadonna committed:
KAFKA-10199: Enable state updater by default

Now that the implementation for the state updater is done, we
can enable it by default.

This PR enables the state updater by default and fixes code that
made assumptions that are not true when the state updater is
enabled (mainly tests).

lucasbru commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1600,7 +1617,21 @@ List<Task> activeTaskIterable() {
         return activeTaskStream().collect(Collectors.toList());
     }
 
+    List<Task> activeRunningTaskIterable() {
```
What does "Running" mean? Is this existing terminology or are you introducing new terms here? What is the difference to "Processing" tasks, and didn't we also have "Owned" tasks somewhere?

cadonna replied:
`RUNNING` is the state in which an active task is when it is processing. 
https://github.com/apache/kafka/blob/750a3893081fbe1e01f8c6eadf84726ace1eb47a/streams/src/main/java/org/apache/kafka/streams/processor/internals/Task.java#L44

cadonna replied:
You are right, there is also `allOwnedTasks()`. All owned tasks are the tasks owned by the stream thread. i.e., all tasks with state updater disabled and only the tasks owned by the stream thread and not by the state updater with state updater enabled. So it is a bit confusing, but luckily only a temporary confusion. A ToDo above `allOwnedTasks()` states:

```java
     * TODO: after we complete switching to state updater, we could rename this function as allRunningTasks
     *       to be differentiated from allTasks including running and restoring tasks
```

cadonna commented on streams/src/main/java/org/apache/kafka/streams/StreamsConfig.java:
```diff
@@ -1171,6 +1171,7 @@ public static class InternalConfig {
 
         // Private API to enable the state updater (i.e. state updating on a dedicated thread)
         public static final String STATE_UPDATER_ENABLED = "__state.updater.enabled__";
+        public static final boolean STATE_UPDATER_ENABLED_DEFAULT = true;
```
Enabling the state updater by default.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -230,8 +230,11 @@ public StoreChangelogReader(final Time time,
 
-        this.stateUpdaterEnabled =
-            InternalConfig.getBoolean(config.originals(), InternalConfig.STATE_UPDATER_ENABLED, false);
+        this.stateUpdaterEnabled = InternalConfig.getBoolean(
+            config.originals(),
+            InternalConfig.STATE_UPDATER_ENABLED,
+            InternalConfig.STATE_UPDATER_ENABLED_DEFAULT
+        );
```
Enabling the state updater by default.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -368,8 +368,11 @@ public static StreamThread create(final TopologyMetadata topologyMetadata,
 
-        final boolean stateUpdaterEnabled =
-            InternalConfig.getBoolean(config.originals(), InternalConfig.STATE_UPDATER_ENABLED, false);
+        final boolean stateUpdaterEnabled = InternalConfig.getBoolean(
+                config.originals(),
+                InternalConfig.STATE_UPDATER_ENABLED,
+                InternalConfig.STATE_UPDATER_ENABLED_DEFAULT
+        );
```
Enabling the state updater by default.

lucasbru replied:
could make sense to make a little `InternalConfig.getStateUpdateEnabled(config)`

cadonna replied:
Good idea!
Done!

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -551,7 +554,11 @@ public StreamThread(final Time time,
         this.numIterations = 1;
         this.eosEnabled = eosEnabled(config);
-        this.stateUpdaterEnabled = InternalConfig.getBoolean(config.originals(), InternalConfig.STATE_UPDATER_ENABLED, false);
+        this.stateUpdaterEnabled = InternalConfig.getBoolean(
+                config.originals(),
+                InternalConfig.STATE_UPDATER_ENABLED,
+                InternalConfig.STATE_UPDATER_ENABLED_DEFAULT
+        );
```
Enabling the state updater by default.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/ReadOnlyTask.java:
```diff
@@ -201,7 +201,7 @@ public boolean commitNeeded() {
     @Override
     public StateStore getStore(final String name) {
-        throw new UnsupportedOperationException("This task is read-only");
+        return task.getStore(name);
```
Needed in some tests

lucasbru replied:
Would be nice to expose a read only state store here (hiding init, flush, close and the like), but that's probably for a differenet PR

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -846,7 +853,7 @@ void runOnce() {
                     if (log.isDebugEnabled()) {
                         log.debug("Committed all active tasks {} and standby tasks {} in {}ms",
-                            taskManager.activeTaskIds(), taskManager.standbyTaskIds(), commitLatency);
+                            taskManager.activeRunningTaskIds(), taskManager.standbyTaskIds(), commitLatency);
```
This is basically a renaming to make the code clearer. With the state updater restoring tasks are owned by the state updater.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1138,28 +1140,35 @@ public void signalResume() {
+            } catch (final IOException e) {
+                log.warn(String.format("Exception caught while trying to read checkpoint for task %s:", id), e);
             }
         }
```
This is the fix from https://github.com/apache/kafka/pull/13925. This will disappear once the fix is merged and this PR rebased.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1542,12 +1551,20 @@ Set<TaskId> activeTaskIds() {
             .collect(Collectors.toSet());
     }
 
+    Set<TaskId> standbyTaskIds() {
+        if (stateUpdater != null) {
+            return stateUpdater.getStandbyTasks().stream().map(Task::id).collect(Collectors.toSet());
+        } else {
+            return standbyTaskStream().map(Task::id).collect(Collectors.toSet());
+        }
+    }
```
With the state updater enabled, standby tasks are owned by the state updater.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1600,7 +1617,21 @@ List<Task> activeTaskIterable() {
+    }
+
     private Stream<Task> activeTaskStream() {
+        if (stateUpdater != null) {
+            return Stream.concat(
+                activeRunningTaskStream(),
+                stateUpdater.getTasks().stream().filter(Task::isActive)
+            );
+        }
```
With the state updater enabled, restoring tasks are owned by the state updater. So they need to be retrieved from there and merged with the running active tasks.

wcarlson5 replied:
Should the state updater have active tasks at all? If so who owns them?

cadonna replied:
Yes, the state updater has active tasks. An active task that needs to be restored is given to the state updater which restores its state. The state updater owns active tasks in state `RESTORING` and standby tasks in state `RUNNING`.

cadonna committed:
Consolidate standby retrieval

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1613,7 +1640,14 @@ private List<Task> standbyTaskIterable() {
 
     private Stream<Task> standbyTaskStream() {
-        return tasks.allTasks().stream().filter(t -> !t.isActive());
+        if (stateUpdater != null) {
+            return Stream.concat(
+                stateUpdater.getStandbyTasks().stream(),
+                tasks.allTasks().stream().filter(t -> !t.isActive())
+            );
+        } else {
```
With enabled state updater, the state updater owns standby tasks. However, when the state updater is shut down it moves its stand by tasks to the stream thread for closing out. Thus, we also need to add `tasks.allTasks().stream().filter(t -> !t.isActive())` here.

cadonna commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1881,15 +1915,11 @@ public static void executeAndMaybeSwallow(final boolean clean,
     void addTask(final Task task) {
         tasks.addTask(task);
     }
-
-    TasksRegistry tasks() {
-        return tasks;
-    }
```
Never used.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/KTableKTableForeignKeyInnerJoinCustomPartitionerIntegrationTest.java:
```diff
@@ -207,11 +207,12 @@ public void shouldThrowIllegalArgumentExceptionWhenCustomPartitionerReturnsMulti
         }
 
-        startApplicationAndWaitUntilRunning(kafkaStreamsList, ofSeconds(120));
+        for (final KafkaStreams stream: kafkaStreamsList) {
+            stream.start();
+        }
```
With the state updater enabled, the Streams clients might still be in `REBALANCING` when the task that throws the `IllegalStateExeption` is processing. Tasks are processed and restored in parallel, but the stream thread transitions to RUNNING when all tasks are restored.

wcarlson5 replied:
That makes sense, but do we still need the synchronization point or even the timeout here?

What I'm asking is is maybe we have a  `waitForApplicationState` for running after calling start for them all

cadonna replied:
I am not sure I understand what you are asking, so bear with me if my reply is not the one you expected. 

Adding
```
waitForApplicationState(Arrays.asList(streams, streamsTwo, streamsThree), KafkaStreams.State.RUNNING, ofSeconds(60))
```
after starting all Streams clients makes the test flaky because with the state updater NOT all clients need to be in `RUNNING` before a Streams client transits to `ERROR`. Not even the Streams client that throws the exception needs to be in `RUNNING` before transitioning to `ERROR`. 
Actually, now that I looked more at this code, I am wondering why this test was not flaky before my changes. Each Streams client is started independently. That means, there is no guarantee, that all three Streams clients are `RUNNING` at the same time. I did some testing with `Thread.sleep()` after each client were started to simulate delays in processing in the version of the test before my changes and the test was flaky indeed.

wcarlson5 replied:
Okay that is very interesting. Good find then!

cadonna committed:
Remove a state updater disablement

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java:
```diff
@@ -113,6 +113,7 @@ public void before(final TestInfo testInfo) {
         streamsConfiguration.put(StreamsConfig.DEFAULT_KEY_SERDE_CLASS_CONFIG, Serdes.String().getClass());
         streamsConfiguration.put(StreamsConfig.DEFAULT_VALUE_SERDE_CLASS_CONFIG, Serdes.String().getClass());
         streamsConfiguration.put(StreamsConfig.COMMIT_INTERVAL_MS_CONFIG, 100L);
+        streamsConfiguration.put(StreamsConfig.InternalConfig.STATE_UPDATER_ENABLED, false);
```
doesn't the integration test work with state updater? A comment would be good here at least

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:
```diff
@@ -173,9 +173,9 @@ public static void closeCluster() {
     private static final List<KeyValue<String, Long>> STANDARD_INPUT_DATA =
         asList(pair("A", 100L), pair("B", 200L), pair("A", 300L), pair("C", 400L), pair("C", -50L));
     private static final List<KeyValue<String, Long>> COUNT_OUTPUT_DATA =
-        asList(pair("B", 1L), pair("A", 2L), pair("C", 2L)); // output of count operation with caching
```
@wcarlson5 @ableegoldman Could you review this changes to know if I changed something I should not have done?

cadonna replied:
I turned off the caching because with the state updater the results were different than without the state updater. The differences were only there due to caching. I think the processing and commit times differ when the state updater is on or off. 
If caching is not needed, I think turning it off makes tests more robust.

wcarlson5 replied:
I don't think that output should have changed. That is worrying

cadonna replied:
@wcarlson5 Could you elaborate what is worrying you here?

I ran the tests with disabled state updater and the disabled cache and all tests pass. That tells me that the state updater does not change any results.
Additionally, the results with enabled and disabled cache are semantically equivalent.
```
pair("B", 1L), pair("A", 2L), pair("C", 2L)
```
is equivalent to
```
pair("A", 1L), pair("B", 1L), pair("A", 2L), pair("C", 1L), pair("C", 2L)
```
The latter just produces more intermediate results.

In general, if a test verifies the correctness of results, disabling the cache makes the test more robust, because the results do not depend on commit time intervals or production rates.

Let me know if I missed something here.

wcarlson5 replied:
Okay that works for me, thanks for the explanation I missed that they were equivalent

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:
```diff
@@ -208,6 +208,8 @@ private Properties configProps(final String appId, final String host) {
         streamsConfiguration.put(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest");
         streamsConfiguration.put(ConsumerConfig.SESSION_TIMEOUT_MS_CONFIG, 10 * 1000);
+        streamsConfiguration.put(StreamsConfig.STATESTORE_CACHE_MAX_BYTES_CONFIG, 0);
+        streamsConfiguration.put(StreamsConfig.InternalConfig.STATE_UPDATER_ENABLED, false);
```
Same here - why do we disable the state updater here?

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/integration/NamedTopologyIntegrationTest.java:
```diff
@@ -173,9 +173,9 @@ public static void closeCluster() {
         asList(pair("A", 100L), pair("B", 200L), pair("A", 300L), pair("C", 400L), pair("C", -50L));
     private static final List<KeyValue<String, Long>> COUNT_OUTPUT_DATA =
-        asList(pair("B", 1L), pair("A", 2L), pair("C", 2L)); // output of count operation with caching
+        asList(pair("A", 1L), pair("B", 1L), pair("A", 2L), pair("C", 1L), pair("C", 2L));
```
why did the output data change?

cadonna replied:
See https://github.com/apache/kafka/pull/13927/files#r1246594593

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -1123,7 +1138,10 @@ public void shouldInjectProducerPerThreadUsingClientSupplierOnCreateIfEosV2Enabl
 
-        final StreamThread thread = createStreamThread(CLIENT_ID, new StreamsConfig(configProps(true)), true);
+        final Properties props = configProps(true);
+        props.put(InternalConfig.STATE_UPDATER_ENABLED, false);
```
same here. if possible, a comment why this test is invalid for state updater

cadonna replied:
Done

lucasbru commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -3038,6 +3149,7 @@ public void shouldRespectPollTimeInPartitionsAssignedStateWithStateUpdater() {
     @Test
     public void shouldNotBlockWhenPollingInPartitionsAssignedStateWithoutStateUpdater() {
         final Properties streamsConfigProps = StreamsTestUtils.getStreamsConfig();
+        streamsConfigProps.put(InternalConfig.STATE_UPDATER_ENABLED, false);
```
In the corresponding SU test, can we remove the `streamsConfigProps.put`?

cadonna replied:
I thought the code is better readable if I leave it in the corresponding with-state-updater test.

cadonna committed:
Remove disable state updater

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java:
```diff
@@ -87,7 +88,7 @@ public static void closeCluster() {
 
 
-    private final Time time = CLUSTER.time;
+    private final Time time = new MockTime(1);
```
The time needs to progress when the state updater is shutdown, otherwise the shutdown is blocked. `CLUSTER` provides a mock time that does not progress.

mjsax replied:
Are we not worried that broker time and our time might diverge? Can we make the broker mock time advance? Seems to be an undesired pattern to have two mock-times?

cadonna replied:
I tried to expose the `autoTickMs` parameter in embedded kafka, but I ran into a infinite loop.

Regarding your question:

> Are we not worried that broker time and our time might diverge?

In almost all our integration tests broker time diverges from streams time since the broker time does not advance but we usually use system time in `KafkaStreams`. This is not something this PR introduces.

cadonna committed:
Add docs for verification methods

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/utils/IntegrationTestUtils.java:
```diff
@@ -843,15 +846,19 @@ private static <K, V, T> List<T> waitUntilFinalKeyValueRecordsReceived(final Pro
-                    finalAccumData.computeIfAbsent(
-                        withTimestamp ? ((KeyValueTimestamp<K, V>) kv).key() : ((KeyValue<K, V>) kv).key,
-                        key -> new ArrayList<>()).add(kv);
+                    final K key = withTimestamp ? ((KeyValueTimestamp<K, V>) kv).key() : ((KeyValue<K, V>) kv).key;
+                    final List<T> records = finalAccumData.computeIfAbsent(key, k -> new ArrayList<>());
+                    if (!records.contains(kv)) {
+                        records.add(kv);
+                    }
                 }
                 final Map<K, List<T>> finalExpected = new HashMap<>();
                 for (final T kv : expectedRecords) {
-                    finalExpected.computeIfAbsent(
-                        withTimestamp ? ((KeyValueTimestamp<K, V>) kv).key() : ((KeyValue<K, V>) kv).key,
-                        key -> new ArrayList<>()).add(kv);
+                    final K key = withTimestamp ? ((KeyValueTimestamp<K, V>) kv).key() : ((KeyValue<K, V>) kv).key;
+                    final List<T> records = finalExpected.computeIfAbsent(key, k -> new ArrayList<>());
+                    if (!records.contains(kv)) {
+                        records.add(kv);
+                    }
```
These changes do not consider duplicate record during the comparison. A test that used this verification triggered a failure and verified then that the expected records were in the output topic. However, without the state updater no records were written to the output topic before the failure. With the state updater some records were written to the output topics.
From a correctness point of view, both is correct since no commit happens before the simulated failure and so Streams reads again all input records after the failover (under ALOS).

cadonna committed:
Add comments why state updater is disabled in tests

cadonna committed:
Add method in config to get whether state updater enabled

lucasbru commented:
Looking good from my side, pending a review from @wcarlson5 or @ableegoldman w.r.t. the caching question.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java:
```diff
@@ -277,6 +277,7 @@ public void shouldFetchLagsDuringRestoration() throws Exception {
         t1.toStream().to(outputTopicName);
         final KafkaStreams streams = new KafkaStreams(builder.build(), props);
```
We need to `close()` a `KafkaStreams` client even if we never `.start()` it, and thus should use `try-with-resource` or split this up and do:
```
final KafkaStreams streams;
try {
  streams = new KafkaStreams(...);
} finally {
  ...
}
```

cadonna replied:
Actually the Streams clients are closed in the `finally`-clause. The thing that was missing was that the Streams clients were not created in the `try`-clause. I needed to use `AtomicReference` for that, otherwise I could not use the Streams clients in lambdas.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java:
```diff
@@ -313,9 +313,14 @@ public void shouldFetchLagsDuringRestoration() throws Exception {
-            // wait till the lag goes down to 0
-            final KafkaStreams restartedStreams = new KafkaStreams(builder.build(), props);
+        // wait till the lag goes down to 0
+        final KafkaStreams restartedStreams = new KafkaStreams(builder.build(), props);
```
as above.

mjsax reviewed (approved):
Also wondering about the caching question... Leave it to your judgement to merge.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -571,7 +571,9 @@ private Map<TaskId, Set<TopicPartition>> pendingTasksToCreate(final Map<TaskId,
             final Map.Entry<TaskId, Set<TopicPartition>> entry = iter.next();
             final TaskId taskId = entry.getKey();
-            if (taskId.topologyName() != null && !topologyMetadata.namedTopologiesView().contains(taskId.topologyName())) {
+            final boolean taskIsOwned = tasks.allTaskIds().contains(taskId)
```
If we are going to just invert this in the next line and not use it anywhere else can we move that inversion here?

cadonna replied:
The idea was to make the code more readable since the condition is quite complex but it actually just verifies that a task is owned. I prefer to use an additional variable or method rather than use complex logical conditions that give the reader a hard time. So, I would rather keep the variable.

wcarlson5 commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java:
```diff
@@ -1619,7 +1639,14 @@ private List<Task> standbyTaskIterable() {
+                tasks.allTasks().stream().filter(t -> !t.isActive())
+            );
+        } else {
+            return tasks.allTasks().stream().filter(t -> !t.isActive());
```
This is the same line right above it. Can we simplify them?

cadonna replied:
Done!

wcarlson5 reviewed (commented):
thanks for the PR!

cadonna committed:
Include feedback

mjsax commented on streams/src/test/java/org/apache/kafka/streams/integration/LagFetchIntegrationTest.java:
```diff
@@ -308,23 +309,30 @@ public void shouldFetchLagsDuringRestoration() throws Exception {
                 .map(Path::toFile)
                 .forEach(f -> assertTrue(f.delete(), "Some state " + f + " could not be deleted"));
+        } finally {
+            streams.get().close();
```
Not sure why we need `AtomicReference`? Can you elaborate?

Why does this not work?
```
KafkaStream streams;
try {
  stream = new KafkaStreams(...);
  ...
```

Btw: `streams.get()` could return `null` in case we fail before calling `set(...)`. Need a `null`-check here.

cadonna replied:
Look at line 284. There `streams` is used in a lambda expression. You can only use variables that are final or effective final in a lambda expression. There are two possibilities to make `streams` final: either we use a final `AtomicReference` or we assign `streams` to an effective final variable and use that in the lambda. I chose the `AtomicReference` option. This are the two options because we need to initialize the variable outside of the `try`-clause to have an initialized variable in the `finally`-clause. 
However, I now realized that we can create the Streams client outside the `try`-clause because if the constructor fails we do not need to call close and cleanup in the `finally`-clause.

cadonna committed:
Get rid of AtomicReference

wcarlson5 reviewed (approved):
LGTM

cadonna committed:
Increase timeout of PurgeRepartitionTopicIntegrationTest

cadonna committed:
Set log level for Streams to DEBUG

cadonna committed:
Set log level for Streams to INFO

cadonna committed:
Fix EosV2UpgradeIntegrationTest

cadonna committed:
Add logging for debugging PurgeRepartitionTopicIntegrationTest

cadonna committed:
Add logging for debugging PurgeRepartitionTopicIntegrationTest

cadonna committed:
Set commit interval to double of purge interval

cadonna committed:
Fix checkstyle error

cadonna committed:
Add log messages and set purge interval to 5ms

cadonna committed:
Add log messages, revert config changes and switch off other tests

cadonna committed:
Add debug  messages for purge future

cadonna committed:
Reduce debug messages to avoid output truncation on jenkins

cadonna committed:
Re-introduce old verification

cadonna committed:
Increase limit for successful test

cadonna committed:
Remove debug messages and try to purge even when not committing

cadonna committed:
Do not reset delete records future

cadonna committed:
Send delete record call when committing

cadonna commented on streams/src/test/java/org/apache/kafka/streams/integration/PurgeRepartitionTopicIntegrationTest.java:
```diff
@@ -212,10 +213,11 @@ public void shouldRestoreState() throws Exception {
-            new RepartitionTopicVerified(currentSize -> currentSize <= PURGE_SEGMENT_BYTES),
-            60000,
-            "Repartition topic " + REPARTITION_TOPIC + " not purged data after 60000 ms."
+            new RepartitionTopicVerified(currentSize -> currentSize <= 4L * PURGE_SEGMENT_BYTES),
```
In the builds on Jenkins, this test failed because a future of a delete record call did not complete before all offsets in the repartition topic got committed. When the delete call finally completes there a no commits that trigger a new delete record call. That means, if the size of the repartition topic after the delete call that did not complete before all commits is still larger than `PURGE_SEGMENT_BYTES`, the tests fails.
So increasing the limit in the test makes it more likely that the test succeeds and at the same time it verifies that most of the segments are deleted.
Alternatively, we could decouple deleting records in the repartition topic from committing. We would need to remove the condition `committed > 0` from this code line: 
https://github.com/apache/kafka/blob/56ab2f80343d7d842c1b0e3069833106d4fe3259/streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java#L1117
We do not need to wait for a commit to delete records in the repartition topic, because if the async call to delete records takes a while to complete, there might be offsets to delete independently of the next commit. All commits since the execution of the delete record call might also have committed offsets that still need to be deleted from the repartition topic. Such a decoupling implies an iteration over all tasks of a stream thread even if no offsets were committed, but it does not imply an additional remote call to delete the records since if no records need to be deleted no remote call is executed.
AFAIK, Kafka Streams only guarantees that records are deleted eventually from the repartition topic, so I opted for the change in the test instead of the change in the hot path of the stream thread.
I tested both options and both options seem to work.
I am not sure why this flakiness was surfaced by enabling the state updater, since it could theoretically also happen with disabled state updater. Maybe the additional state updater thread limits the resources available to the embedded Kafka. But also that seems unlikely since the state updater thread should be waiting on a condition variable at that stage in the test.

cadonna replied:
@lucasbru @mjsax @wcarlson5 @aliehsaeedii Could you please sanity check this single change?

lucasbru replied:
For my understanding: why can there only be one delete call in-flight at the same time? Where is this prevented?

cadonna replied:
It is prevented here: https://github.com/apache/kafka/blob/7782741262c08e5735f7c8e09727ec37cb5f7f02/streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java#L1782

I do not know why only one delete call should be in-flight at the same time. Maybe to limit the load of admin calls on the brokers?

lucasbru replied:
Yeah, that makes sense. Seems to me that the overhead wouldn't be noticable if we execute the code even if we didn't commit anything. And it would make the execution of streams slightly less surprising, so, that's what we are going for right? I vote for purging no matter what.

cadonna replied:
I opened the following PR: https://github.com/apache/kafka/pull/14227

cadonna replied:
Merged https://github.com/apache/kafka/pull/14227

cadonna committed:
Wait for RUNNING in NamedTopologyIntegrationTest

cadonna committed:
Improve error message if not all Streams clients reach the RUNNING state

cadonna committed:
KAFKA-10199: Add standbys recycled from running tasks to state updater

We missed to add standby tasks recycled from running active tasks to the
state updater. This caused IllegalStateExceptions in the following
rebalance and shutdowns of Kafka Streams clients to state ERROR.

This commit adds standby tasks recycled from running active tasks to
the pending tasks to init. Those tasks are eventually added to the
state updater. Additionally, the running active task is removed from
the task registry.

cadonna committed:
Reduce expected size of purged partitions in test

cadonna committed:
Fix merge mistake

lucasbru reviewed (approved):
lgtm
